### PR TITLE
fix: allow range headers to be set in getObject request when offset is 0 (#706)

### LINKED ISSUE #706: Unable to get first X bytes of object
I am unable to get the first X bytes of object via getObject(String bucketName, String objectName, long offset, Long length). Specifying a 0 offset and X length would return an entire file rather than bytes from 0 - X-1

Directly Quoting RFC 2616 Page 138, Regarding Byte range.

> The first 500 bytes (byte offsets 0-499, inclusive):  bytes=0-499

Further inspection of the source code, I found that if offset is less than or equal to 0, range header would get ignored.
Screenshot of source code from `MinioClient.java` Version 5.0.1
screen shot 2018-10-01 at 02 17 46

### PR DESCRIPTION
**Pull request by christopher-fredregill**
fix: allow range headers to be set in getObject request when offset is 0 (#706)

fixes #706

### CODE REVIEW

**Comment by christopher-fredregill** [comment]
Some notes for anyone reviewing this:

1. The test that's failing in Travis does not fail when run against a minio host on RELEASE.2018-08-02T23-11-36Z; it seems like there is some sort of regression in whatever version of minio is currently backing the https://play.minio.io:9000 endpoint.

Here's a docker-compose snippet you can use to spin up a local test endpoint against which the full functional test suite (on my feature branch) will pass:

```
version: '3.7'
services:
  minio:
    image: minio/minio:RELEASE.2018-08-02T23-11-36Z
    command: server /data
    ports:
      - 9000:9000
    volumes:
      - ./data/:/data/
```

2. Please see https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L1627 in the original implementation. When the length is `null`, but the offset is valid, something like `Range: bytes=X-` (where `X` is the user-supplied offset) gets added to the request headers. This always results in a `416/The requested range is not satisfiable` error response, in the test suite. You'll notice that, as a workaround, I've placed the failing test in a try/catch block, but I am wondering if there is any reason why we should not simply require that a valid length be given. Please advise as to how best to handle this.

3. None of the test cases I found when I started looking into this problem were actually capable of detecting the error, so I modified one of the functional test cases to make some stronger assertions about the response from a partial getObject request. I'm open to other suggestions around how to verify the functionality in question.

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -1620,10 +1620,11 @@ public InputStream getObject(String bucketName, String objectName, long offset,
     }
 
     Map<String,String> headerMap = new HashMap<>();
-    if (offset > 0) {
+    if (offset >= 0) {
       if (length != null) {
         headerMap.put("Range", "bytes=" + offset + "-" + (offset + length - 1));
       } else {
+        // results in a 416 with response message "The requested range is not satisfiable"
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: This comment is not needed

File: functional/FunctionalTest.java [hunk 2]
```diff
@@ -1033,12 +1033,24 @@ public static void getObject_test3() throws Exception {
     long startTime = System.currentTimeMillis();
     try {
       String objectName = getRandomName();
-      try (final InputStream is = new ContentInputStream(3 * MB)) {
-        client.putObject(bucketName, objectName, is, 3 * MB, nullContentType);
+      try (final InputStream is = new ContentInputStream(1024)) {
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Add new tests than changing existing ones.

File: api/src/main/java/io/minio/MinioClient.java [hunk 3]
```diff
@@ -1620,10 +1620,11 @@ public InputStream getObject(String bucketName, String objectName, long offset,
     }
 
     Map<String,String> headerMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java]: getObject() in minio-java does accept offset and length not range i.e. not start/end values.  With this existing fix is correct.

[the remark this answers is not in the document]
Comment (harshavardhana) [reply api/src/main/java/io/minio/MinioClient.java]: There are three allowed behaviors

- bytes=N-M
- bytes=N-
- bytes=-M 

In minio-java the last one is not supported this is what I meant by complete implementation @balamurugana

File: functional/FunctionalTest.java (hunk 2 above)
Comment (christopher-fredregill) [reply functional/FunctionalTest.java]: I've moved the new test case (with the assertions about the bytes that should be received from the partial getObject request) into `getObject_test8` in the functional suite. I had to make minor updates to two other functional tests that were leaving behind objects in the shared bucket and causing `listObject_test5` to fail in error; those other tests should now be cleaning up after themselves.

**Comment by christopher-fredregill** [comment]
@harshavardhana @balamurugana can you please take another look?

File: functional/FunctionalTest.java [hunk 4]
```diff
@@ -1125,8 +1127,14 @@ public static void getObject_test6() throws Exception {
         client.putObject(bucketName, objectName, is, 0, nullContentType);
       }
 
-      client.getObject(bucketName, objectName)
-          .close();
+      try {
+        client.getObject(bucketName, objectName).close();
+      } catch (final ErrorResponseException e) {
+        // current implementation results in request headers="Range: bytes=0-"; response is a 416 with the given message
```
Comment (harshavardhana) [inline functional/FunctionalTest.java]: We can remove this comment, this is an expected response.

Comment (christopher-fredregill) [reply functional/FunctionalTest.java]: removed

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -1620,12 +1620,10 @@ public InputStream getObject(String bucketName, String objectName, long offset,
...
-    if (offset > 0) {
-      if (length != null) {
-        headerMap.put("Range", "bytes=" + offset + "-" + (offset + length - 1));
-      } else {
-        headerMap.put("Range", "bytes=" + offset + "-");
-      }
+    if (length != null) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: `if (offset > 0)` has to be there else wrong range will be sent for negative values.

File: functional/FunctionalTest.java [hunk 6]
```diff
@@ -1125,8 +1127,13 @@ public static void getObject_test6() throws Exception {
...
 
-      client.getObject(bucketName, objectName)
-          .close();
+      try {
+        client.getObject(bucketName, objectName).close();
+      } catch (final ErrorResponseException e) {
+        if (!e.toString().contains("message=The requested range is not satisfiable")) {
+          throw e;
+        }
+      }
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Unable to understand this change.  Why is it required for this test?

File: functional/FunctionalTest.java [hunk 7]
```diff
@@ -1187,6 +1194,63 @@ public static void getObject_test7() throws Exception {
     }
   }
 
+  /**
+   * Test: getObject(String bucketName, String objectName, long offset, Long length).
+   */
+  public static void getObject_test8() throws Exception {
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: You would need to add specific to `offset=0`.  `getObject_test2()` and getObject_test3()` are covering other cases.  Additionally keep only one test in test method.

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (christopher-fredregill) [reply api/src/main/java/io/minio/MinioClient.java]: The linter was complaining in CI because this check is actually unnecessary; if you look a few lines up, there is already a check for the case when offset < 0. I had to remove the duplicate check here in order to satisfy the linter.

File: functional/FunctionalTest.java (hunk 6 above)
Comment (christopher-fredregill) [reply functional/FunctionalTest.java]: The original test catches and re-throws any exception it encounters.

I modified the behavior of the client, such that a known 416 exception will always be encountered here.

The test now has to know in advance that it will see this kind of exception and accept it, or else it will simply fail.

This is why I included comments explaining this in both the client code and in the fictional test code.

Some sort of change is necessary here... maybe you can propose an alternative solution, if this doesn’t suit you.

Comment (harshavardhana) [reply functional/FunctionalTest.java]: The problem is we send offset requests by default now - after this PR change even for empty objects which is not allowed in S3 

This exception check is needed, or we make sure in this test Range is not set for empty objects.

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java]: got it

File: api/src/main/java/io/minio/MinioClient.java [hunk 8]
```diff
@@ -1620,12 +1620,10 @@ public InputStream getObject(String bucketName, String objectName, long offset,
...
+      headerMap.put("Range", "bytes=" + offset + "-" + (offset + length - 1));
+    } else {
+      headerMap.put("Range", "bytes=" + offset + "-");
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: To be clean you would need to set `Range` for non-zero offset

Comment (christopher-fredregill) [reply api/src/main/java/io/minio/MinioClient.java]: Not sure what has to change, specifically; could you clarify what you'd like the `Range` to look like, in each of the scenarios under consideration? (`offset`=0 vs. `offset` > 0, with and without positive `length` argument, etc.).

Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java]: You would need to do below;
```java
if (length != null) {
  headerMap.put("Range", "bytes=" + offset + "-" + (offset + length -1));
  // Here offset is set anyway to satisfy range i.e. `Range: bytes=0-100` or `Range: bytes=10-100`
} else if offset > 0 {
  headerMap.put("Range", "bytes=" + offset + "-");
  // Here `Range` header is set only if offset > 0 i.e. for offset=10; `Range: bytes=10-`; for offset=0; nothing is set.
}
```

Comment (christopher-fredregill) [reply api/src/main/java/io/minio/MinioClient.java]: Updated the client and test code as per ^. Thanks.

**Comment by harshavardhana** [comment]
We could add a test to generate the 416 range not satisfiable with offset=1 for a zero byte object @christopher-fredregill upto you though. This PR is okay to be merged.
